### ISSUE TITLE
False positive for operator-assignment for bitwise operators

### ISSUE THREAD

**Issue by webfrontzhifei** [eslint/eslint#9639]
Code:   status = 1 ^ status;
It throw errors " Assignment can be replaced with operator assignment"

**Comment by platinumazure** [comment]
This is a bug. I think it's because we're special-casing operators that we know are non-commutative (e.g., `-`), but we should really go the other way around and special-case the commutative operators.

Seems this affects bitwise and (`&`), bitwise or (`|`), and bitwise not (`^`). Reproduction [here](https://eslint.org/demo/#eyJ0ZXh0IjoiLyogZXNsaW50IG9wZXJhdG9yLWFzc2lnbm1lbnQ6IGVycm9yICovXHJcblxyXG4vLyBjb3JyZWN0IChubyB3YXJuaW5nKTpcclxuc3RhdHVzID0gXCJcIiArIHN0YXR1cztcclxuc3RhdHVzID0gMSArIHN0YXR1cztcclxuc3RhdHVzID0gMSAtIHN0YXR1cztcclxuc3RhdHVzID0gMSAvIHN0YXR1cztcclxuc3RhdHVzID0gMSAlIHN0YXR1cztcclxuc3RhdHVzID0gMSA8PCBzdGF0dXM7XHJcbnN0YXR1cyA9IDEgPj4gc3RhdHVzO1xyXG5zdGF0dXMgPSAxID4+PiBzdGF0dXM7XHJcbnN0YXR1cyA9IDEgKiogc3RhdHVzO1xyXG5cclxuLy8gY29ycmVjdCAod2FybmluZyk6XHJcbnN0YXR1cyA9IDEgKiBzdGF0dXM7XHJcblxyXG4vLyBpbmNvcnJlY3QgKHdhcm5pbmcpOlxyXG5zdGF0dXMgPSAxIF4gc3RhdHVzO1xyXG5zdGF0dXMgPSAxICYgc3RhdHVzO1xyXG5zdGF0dXMgPSAxIHwgc3RhdHVzOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6OCwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==).

**Comment by not-an-aardvark** [comment]
I don't understand, it seems like the error in that case is correct because the code could be replaced with

```js
status ^= 1;
status &= 1;
status |= 1;
```

**Comment by platinumazure** [comment]
I thought `status ^= 1` was equivalent to `status = status ^ 1` rather than
`status = 1 ^ status`. Am I mistaken?

**Comment by not-an-aardvark** [comment]
You're correct that it's equivalent to `status = status ^ 1`, but I think the `^` operator is commutative, so it's also equivalent to `status = 1 ^ status`.

**Comment by platinumazure** [comment]
Fiddlesticks, you're right. That applies to all three.

In that case, closing due to lack of response.
